src/test/regress/upgrade/upgrade_test.py

```diff
@@ -0,0 +1,181 @@
+#!/usr/bin/env python3
+
+"""upgrade_test
+Usage:
+    upgrade_test --old-bindir=<old-bindir> --new-bindir=<new-bindir> --pgxsdir=<pgxsdir>
+
+Options:
+    --old-bindir=<old-bindir>              The old PostgreSQL executable directory;(ex: '~/.pgenv/pgsql/bin')
+    --new-bindir=<new-bindir>              New postgres binary absolute path(ex: '~/.pgenv/pgsql-11.3/bin')
+    --pgxsdir=<pgxsdir>           	       Path to the PGXS directory(ex: ~/.pgenv/src/postgresql-11.3)
+"""
+
+import utils
+import atexit
+import subprocess
+import sys
+import shutil
+import os
+
+from docopt import docopt
+
+from config import (
+    Config, USER, NODE_PORTS,
+    NODE_NAMES, DBNAME, COORDINATOR_NAME,
+    WORKER_PORTS, AFTER_UPGRADE_SCHEDULE, BEFORE_UPGRADE_SCHEDULE
+)
+
+
+def initialize_temp_dir(temp_dir):
+    if os.path.exists(temp_dir):
+        shutil.rmtree(temp_dir)
+    os.mkdir(temp_dir)
+    # Give full access to TEMP_DIR so that postgres user can use it.
+    os.chmod(temp_dir, 0o777)
+
+
+def initialize_db_for_cluster(pg_path, rel_data_path, settings):
+    subprocess.call(['mkdir', rel_data_path])
+    for node_name in NODE_NAMES:
+        abs_data_path = os.path.abspath(os.path.join(rel_data_path, node_name))
+        command = [
+            os.path.join(pg_path, 'initdb'),
+            '--pgdata', abs_data_path,
+            '--username', USER
+        ]
+        subprocess.call(command)
+        add_settings(abs_data_path, settings)
+
+
+def add_settings(abs_data_path, settings):
+    conf_path = os.path.join(abs_data_path, 'postgresql.conf')
+    with open(conf_path, 'a') as conf_file:
+        for setting_key, setting_val in settings.items():
+            setting = "{setting_key} = \'{setting_val}\'\n".format(
+                setting_key=setting_key,
+                setting_val=setting_val)
+            conf_file.write(setting)
+
+
+def start_databases(pg_path, rel_data_path):
+    for node_name in NODE_NAMES:
+        abs_data_path = os.path.abspath(os.path.join(rel_data_path, node_name))
+        command = [
+            os.path.join(pg_path, 'pg_ctl'),
+            '--pgdata', abs_data_path,
+            '-U', USER,
+            "-o '-p {}'".format(NODE_PORTS[node_name]),
+            '--log', os.path.join(pg_path, 'logfile_' + node_name),
+            'start'
```

Comment:
nit; I thought this one could also be at the beginning `pg_ctl start ...` might be slightly more readable if that is possible.